### PR TITLE
mt8195 dsp ipc support

### PR DESCRIPTION
**Pull request by kuanhsuncheng**
This patch adds IPC support for mt8195 using mbox.

On mt8195 resource, there are two mboxes are used to handle
ipc request and reply.
We create a mtk-adsp-ipc client device to request mbox
controllers.

Here is the upstrream mail list link.
https://patchwork.kernel.org/project/linux-mediatek/list/?series=605114
https://patchwork.kernel.org/project/linux-mediatek/list/?series=605115

Those still wait for maintainer reviewing.

Please let me know if there is any problem. thanks.

### CODE REVIEW

File: drivers/firmware/mediatek/mtk-adsp-ipc.c [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+int mtk_adsp_ipc_send(struct mtk_adsp_ipc *ipc, unsigned int idx, uint32_t msg)
+{
+	struct mtk_adsp_chan *dsp_chan;
+	int ret;
+
+	if (idx >= MTK_ADSP_MBOX_NUM)
+		return -EINVAL;
+
+	dsp_chan = &ipc->chans[idx];
+	ret = mbox_send_message(dsp_chan->ch, &msg);
```
Comment (cujomalainey) [inline drivers/firmware/mediatek/mtk-adsp-ipc.c]: if you initialize `ret = 0` you can just return here and remove the if statement

File: drivers/firmware/mediatek/mtk-adsp-ipc.c [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	struct mtk_adsp_chan *chan = container_of(c, struct mtk_adsp_chan, cl);
+
+	switch (chan->idx) {
+	case MTK_ADSP_MBOX_REPLY:
+		chan->ipc->ops->handle_reply(chan->ipc);
+		break;
+	case MTK_ADSP_MBOX_REQUEST:
+		chan->ipc->ops->handle_request(chan->ipc);
+		break;
+	default:
```
Comment (cujomalainey) [inline drivers/firmware/mediatek/mtk-adsp-ipc.c:67]: not even a debug trace for a bad message?

**Comment by plbossart** [comment]
> This patch adds IPC support for mt8195 using mbox.
> 
> On mt8195 resource, there are two mboxes are used to handle ipc request and reply. We create a mtk-adsp-ipc client device to request mbox controllers.
> 
> Here is the upstrream mail list link. https://patchwork.kernel.org/project/linux-mediatek/list/?series=589151
> 
> Those still wait for maintainer reviewing.
> 
> Please let me know if there is any problem. thanks.

I am not sure what the ask is here. If those patches have not been accepted by the maintainers just yet, why should we merge them in the SOF tree?

**Comment by kuanhsuncheng** [comment]
> his patch adds IPC support for mt8195 using mbo



> 

The thing bother me is I don't get the reply for maintainers a couple of months.

And those patches is only used for SOF.

We have some other patches for MT8195 needed to merge depending on sof ipc.

I am wonder it's ok those patch can merge them to sof_dev?
 
Please kindly give me advice, thanks.

File: drivers/firmware/mediatek/mtk-adsp-ipc.c (hunk 1 above)
Comment (kuanhsuncheng) [reply drivers/firmware/mediatek/mtk-adsp-ipc.c]: Hi Curtis, 

Sorry, I am not sure what you means. 

If  I just return here using mbox_send_message, it seems no need res variable.

And mbox_send_message will return non-negative integer for successful submission (non-blocking mode) or transmission over chan (blocking mode).

In the above situations, we will return 0.

I think I can't just return  mbox_send_message.

Comment (cujomalainey) [reply drivers/firmware/mediatek/mtk-adsp-ipc.c]: ah so it returns > 0 on success @plbossart FYI as i saw you had the same comment on the message board

Comment (plbossart) [reply drivers/firmware/mediatek/mtk-adsp-ipc.c]: ok, then you definitively need a comment to make sure the handling of positive values is captured

Comment (kuanhsuncheng) [reply drivers/firmware/mediatek/mtk-adsp-ipc.c]: I add a comment to describe this.

File: drivers/mailbox/mtk-adsp-mailbox.c [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	if (!mbox->chans)
+		return -ENOMEM;
+
+	priv->va_mboxreg = devm_platform_ioremap_resource(pdev, 0);
+	if (IS_ERR(priv->va_mboxreg))
+		return PTR_ERR(priv->va_mboxreg);
+
+	cfg = of_device_get_match_data(dev);
+	if (!cfg)
+		return -EINVAL;
```
Comment (ujfalusi) [inline drivers/mailbox/mtk-adsp-mailbox.c]: `of_device_get_match_data(dev)` will never going to return NULL for you as the driver is only going to probe with compatible `mediatek,mt8195-adsp-mbox` and that has the data.
```c
priv->cfg = of_device_get_match_data(dev);
```
is simpler, unless you will have platforms where you will not have the cfg coming via match data.

**Comment by plbossart** [comment]
@kuanhsuncheng this series has been discussed upstream, can we close this PR which hasn't been updated for 4 months?

**Comment by kuanhsuncheng** [comment]
@plbossart 

Hi Pierre,

We can close this PR. 

Mark has applied to  https://urldefense.com/v3/__https:/git.kernel.org/pub/scm/linux/kernel/git/broonie/sound.git__;!!CTRNKA9wMg0ARbw!wU0fgArK_q_tKNK7UqFxsnd2931eAre0YciHrlAxzzQaER8C2tTATfYaiNuALVi85laWXg$  for-next

[1/3] firmware: mediatek: Add adsp ipc protocol interface
      commit: 9db69df4bdd37eb1f65b6931ee067fb15b9a4d5c
[2/3] ASoC: SOF: mediatek: Add ipc support for mt8195
      commit: aea9350108ed1627f8610c93de44578162b3ee91
[3/3] ASoC: SOF: mediatek: Add mt8186 ipc support
      commit: e0100bfd383c7d994d2e957e85ca56a5fe5a3f43

Thanks.

**Pull request closed without merging**